index.bs

```diff
@@ -1736,6 +2079,230 @@ The <dfn method for=StorageManager>getDirectory()</dfn> method steps are:
+
+1. [=Enqueue the following steps=] to the [=file system queue=]:
+  1. If |observerRegistrationMap|[|locator|] [=map/exists=], abort these steps.
+
+  1. Let |entry| be the result of [=locating an entry=] given |locator|.
+  1. Let |accessResult| be the result of running |entry|'s
+     [=file system entry/query access=] given "`read`".
+
+  1. [=Queue a storage task=] with |global| to run these steps:
+    1. If |accessResult|'s [=file system access result/permission state=]
+       is not "{{PermissionState/granted}}", [=/reject=] |result| with a
+       {{DOMException}} of |accessResult|'s
+       [=file system access result/error name=] and abort these steps.
+
+    1. If |entry| is null, [=/reject=] |result| with a
+       "{{NotFoundError}}" {{DOMException}} and abort these steps.
+
+    1. [=Enqueue the following steps=] to the [=file system queue=]:
+      1. [=file system observer registration/Create an observer registration=]
+        for [=this=] on |handle| with |recursive|.
+
+    1.  [=/Resolve=] |result| with null.
+
+</div>
+
+
+### The {{FileSystemObserver/unobserve()}} method ### {#api-filesystemobserver-unobserve}
+
+<div algorithm>
+The <dfn method for=FileSystemObserver>unobserve(FileSystemHandle |handle|)</dfn>
+steps are:
+
+1. Let |locator| be |handle|'s [=FileSystemHandle/locator=].
+1. Let |observerRegistrationMap| be [=this=]'s [=FileSystemObserver/observerRegistrations=].
+
+1. [=Enqueue the following steps=] to the [=file system queue=]:
+  1. If |observerRegistrationMap|[|locator|] does not [=map/exist=], abort these steps.
+  1. [=file system observer registration/Destroy=] |observerRegistrationMap|[|locator|].
+
+</div>
+
+### The {{FileSystemObserver/disconnect()}} method ### {#api-filesystemobserver-disconnect}
+
+<div algorithm>
+The <dfn method for=FileSystemObserver>disconnect()</dfn> steps are:
+
+1. Let |observerRegistrationMap| be [=this=]'s [=FileSystemObserver/observerRegistrations=].
+
+1. [=Enqueue the following steps=] to the [=file system queue=]:
+  1. Let |observerRegistrations| be the result of [=map/getting the values=] of
+     |observerRegistrationMap|.
+  1. [=list/For each=] |observerRegistration| in |observerRegistrations|:
+    1. [=file system observer registration/Destroy=] |observerRegistration|.
+
+</div>
+
+## The {{FileSystemChangeRecord}} interface ## {#api-filesystemchangerecord}
+
+<xmp class=idl>
+enum FileSystemChangeType {
+  "appeared",
+  "disappeared",
+  "errored",
+  "modified",
+  "moved",
+  "unknown",
+};
+
+dictionary FileSystemChangeRecordInit {
+  required FileSystemHandle root;
+  required FileSystemHandle changedHandle;
+  required sequence<USVString> relativePathComponents;
+  required FileSystemChangeType type;
+  sequence<USVString>? relativePathMovedFrom;
+};
+
+[
+    Exposed=(DedicatedWorker,SharedWorker,Window),
+    SecureContext,
+    RuntimeEnabled=FileSystemObserver
+] interface FileSystemChangeRecord {
```

Comment:
Is there a reason this is a class and not just a dictionary? I think in retrospect DOM's `MutationRecord` could have just been a dictionary as well. It's also not entirely clear to me why this has a constructor.

Reply:
I asked the same question, and we had some offline discussion about this. The question of dictionary vs class seems to be the oldest open tag guidance request, but the most recent thinking in the issue thread there (https://github.com/w3ctag/design-principles/issues/11#issuecomment-1515585583) seems to lean arguably in favor of using a class here, or at least not arguing against it (but that guidance certainly has changed a couple of times as the make-up of the tag has changed). I'm not sure what the latest thinking in this area is, and it would be nice if the TAG would actually write up something about this other than in that very long issue thread...

Reply:
I'm not sure that accurately reflects the upthread discussion though with input from me and @domenic. In particular @domenic made the point there that for observer-related cases where it's just about data a dictionary would be more idiomatic. Having both a class and a constructor for it that takes a dictionary seems quite bloated.

Reply:
Per https://groups.google.com/a/chromium.org/g/blink-dev/c/6oOaFmia2dc/m/gx0KgpQqBQAJ, the latest plan seems to be to use a dictionary, especially since that is easier to upgrade to an interface later. I think that's a good plan, and updating the spec PR to match that would be good.

Reply:
Switch to dictionary.